clang/test/SemaSYCL/unnamed-kernel.cpp

```diff
@@ -0,0 +1,84 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl-is-device -fsycl-int-header=%t.h -fsyntax-only -verify %s
+// RUN: %clang_cc1 -I %S/Inputs -fsycl-is-device -fsycl-int-header=%t.h -fsycl-unnamed-lambda -fsyntax-only -verify %s
+#include <sycl.hpp>
+
+#ifdef __SYCL_UNNAMED_LAMBDA__
+// expected-no-diagnostics
+#endif
+
+namespace namespace1 {
+  template <typename T> class KernelName;
+}
+
+struct MyWrapper {
+private:
+  class InvalidKernelName0 {};
+  class InvalidKernelName3 {};
+  class InvalidKernelName4 {};
+  class InvalidKernelName5 {};
+
+public:
+  void test() {
+    cl::sycl::queue q;
+#ifndef __SYCL_UNNAMED_LAMBDA__
+    // expected-error@+2 {{kernel needs to have a globally-visible name}}
+#endif
+    class InvalidKernelName1 {};
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<InvalidKernelName1>([] {});
+    });
+
+#ifndef __SYCL_UNNAMED_LAMBDA__
+    // expected-error@+2 {{kernel needs to have a globally-visible name}}
+#endif
+    class InvalidKernelName2 {};
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<namespace1::KernelName<InvalidKernelName2>>([] {});
+    });
+
+#ifndef __SYCL_UNNAMED_LAMBDA__
+    // expected-error@15 {{kernel needs to have a globally-visible name}}
+#endif
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<InvalidKernelName0>([] {});
+    });
+
+#ifndef __SYCL_UNNAMED_LAMBDA__
+    // expected-error@16 {{kernel needs to have a globally-visible name}}
```

Comment:
I suggest to make this diagnostic a bit more informative. It should point to kernel-name class definitions and defective kernel invoke call both.

Reply:
Done